### PR TITLE
Adding open-source code for the PRECISION paper

### PR DESCRIPTION
**Pull request by chenxiaoqino**
Submitting P4 code accompanying our recent paper _Efficient Measurement on Programmable
Switches Using Probabilistic Recirculation_, to appear in ICNP'18.

### CODE REVIEW

File: research_projects/PRECISION/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Reference Implementation of the PRECISION heavy-hitter algorithm
+
+This dirctory hosts a reference implementation of PRECISION algorithm running on the BMV2 v1model.
+
+
+
+## How to run
+TODO: we may want to adopt the makefile from P4 tutorials to make running experiment easier (e.g. `make run`).
+
+This example uses SrcIP/DstIP pair as flow ID, but you're more than welcomed to change to other flow ID definition.
```
Comment (mhira1) [inline research_projects/PRECISION/README.md]: Typo: "welcomed" -> "welcome"

File: research_projects/PRECISION/LICENSE [hunk 2]
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
```
Comment (mhira1) [inline research_projects/PRECISION/LICENSE]: Please move this LICENSE file under src

File: research_projects/PRECISION/src/precision.p4 [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+control MyVerifyChecksum(inout headers hdr, inout custom_metadata_t meta) {
+    apply {  }
+}
+
+
+control MyIngress(inout headers hdr,
+                  inout custom_metadata_t meta,
+                  inout standard_metadata_t standard_metadata) {
+    action repeater () {
+         standard_metadata.egress_spec = 9w3 - standard_metadata.ingress_port;
```
Comment (mhira1) [inline research_projects/PRECISION/src/precision.p4]: Please add a comment here. What is the expression on the right hand side doing?

File: research_projects/PRECISION/src/entries_naive.txt [hunk 4]
```diff
@@ -0,0 +1,63 @@
+			(64w0b0000000000000000000000000000000000000000000000000000000000000011 &&& 64w0b1111111111111111111111111111111111111111111111111111111111111110, 64w0 &&& 64w0b0000000000000000000000000000000000000000000000000000000000000011 ) : clone_and_recirc_replace_entry();
```
Comment (mhira1) [inline research_projects/PRECISION/src/entries_naive.txt]: .txt sounds unusual for an include file. Can you name it as .h or .p4?

File: research_projects/PRECISION/src/entries_better.txt [hunk 5]
```diff
@@ -0,0 +1,561 @@
+
```
Comment (mhira1) [inline research_projects/PRECISION/src/entries_better.txt]: Can you rename to .h or .p4?

**Comment by chenxiaoqino** [comment]
@mhira1 Thanks for the comments, I've fixed the files accordingly.

**Comment by chenxiaoqino** [comment]
Done adding paper link and citation information

**Review by mhira1 — COMMENTED** [review]
Danny, this is good to commit to the repository. However, can you please squash all 11 commits into 1 commit?

**Comment by chenxiaoqino** [comment]
Hi Mukesh,

Sure, let me figure out how.

**Comment by chenxiaoqino** [comment]
I have squashed all changes into one commit. Can you check if that's been done correctly? Thanks!

**Review by mhira1 — APPROVED** [review]
Thanks for squashing into one commit. Looks good.

**Review by jnfoster — CHANGES_REQUESTED** [review]
Please remove the license file. All contributions to P4 are done under the Apache2 license, per your signed member agreement.

**Comment by jnfoster** [comment]
@chenxiaoqino: just to confirm since I am traveling, are you or your employer a P4 member?

**Comment by chenxiaoqino** [comment]
I believe yes. Although not all authors of the paper are, I myself am
affiliated with a P4 consortium member. I write and open-source the code on
my own behalf.

**Comment by chenxiaoqino** [comment]
@jnfoster Done.

**Comment by jklr** [comment]
@chenxiaoqino  how about adding your name and contact email into the README?

**Comment by chenxiaoqino** [comment]
@jklr Sure

**Review by jklr — APPROVED** [review]

**Pull request merged**